geocode-offline/src/main/java/com/esri/arcgismaps/sample/geocodeoffline/MainActivity.kt

```diff
@@ -0,0 +1,267 @@
+
+    /**
+     * Sets up the address SearchView and uses MatrixCursor to
+     * show suggestions to the user as text is entered.
+     */
+    private fun setupAddressSearchView(addressSearchView: SearchView) {
+        // get the list of pre-made suggestions
+        val suggestions = resources.getStringArray(R.array.suggestion_items)
+        // set up parameters for searching with MatrixCursor
+        val columnNames = arrayOf(BaseColumns._ID, "address")
+        val suggestionsCursor = MatrixCursor(columnNames)
+        // add each address suggestion to a new row
+        suggestions.forEachIndexed { i, s -> suggestionsCursor.addRow(arrayOf(i, s)) }
+
+        // column names for the adapter to look at when mapping data
+        val cols = arrayOf("address")
+        // ids that show where data should be assigned in the layout
+        val to = intArrayOf(R.id.suggestionAddress)
+        // define SimpleCursorAdapter
+        val suggestionsAdapter = SimpleCursorAdapter(
+            this@MainActivity,
+            R.layout.suggestion_address, suggestionsCursor, cols, to, 0
+        )
+
+        addressSearchView.suggestionsAdapter = suggestionsAdapter
+        // handle an address suggestion being chosen
+        addressSearchView.setOnSuggestionListener(object : SearchView.OnSuggestionListener {
+            override fun onSuggestionSelect(position: Int): Boolean {
+                return false
+            }
+
+            override fun onSuggestionClick(position: Int): Boolean {
+                // geocode the typed address
+                geocodeAddress(suggestions[position])
+                addressSearchView.clearFocus()
+                return true
+            }
+        })
+
+        // geocode the searched address on submit
+        addressSearchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
+            override fun onQueryTextSubmit(address: String): Boolean {
+                geocodeAddress(address)
+                addressSearchView.clearFocus()
+                return true
+            }
+
+            override fun onQueryTextChange(newText: String?) = true
+        })
+    }
+
+    /**
+     * Use the locator task to geocode the given address.
+     *
+     * @param address as a string to geocode
+     */
+    private fun geocodeAddress(address: String) = lifecycleScope.launch {
+        // clear graphics on map before displaying search results
+        graphicsOverlay.graphics.clear()
+        // load the locator task
+        locatorTask.load().getOrThrow()
+        // run the locatorTask geocode task, passing in the address
+        val geocodeResults = locatorTask.geocode(address, geocodeParameters).getOrThrow()
+        // no address found in geocode so return
+        if (geocodeResults.isEmpty()) {
+            showError("No address found for $address")
+            return@launch
+        }
+        // display address found in geocode
+        else displaySearchResultOnMap(geocodeResults)
```

Comment:
This might be a bit easier with `getOrElse()`
```suggestion
        val geocodeResults = locatorTask.geocode(address, geocodeParameters).getOrElse {
        // no address found in geocode so return
            showError("No address found for $address")
            return@launch
        }
        // display address found in geocode
        displaySearchResultOnMap(geocodeResults)
```

Reply:
good suggestion @hud10837 I looked through the definition and usages of both and get why getOrElse would be suitable here.

Reply:
good suggestion @hud10837 I looked through the definition and usages of both and get why getOrElse would be suitable here.